runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java

```diff
@@ -978,12 +978,16 @@ public void translate(ParDo.MultiOutput transform, TranslationContext context) {
             if (context.isFnApi()) {
               DoFnSignature signature = DoFnSignatures.signatureForDoFn(transform.getFn());
               if (signature.processElement().isSplittable()) {
-                Coder<?> restrictionCoder =
-                    DoFnInvokers.invokerFor(transform.getFn())
-                        .invokeGetRestrictionCoder(
-                            context.getInput(transform).getPipeline().getCoderRegistry());
+                DoFnInvoker<?, ?> doFnInvoker = DoFnInvokers.invokerFor(transform.getFn());
+                Coder<?> restrictionAndWatermarkStateCoder =
+                    KvCoder.of(
+                        doFnInvoker.invokeGetRestrictionCoder(
+                            context.getInput(transform).getPipeline().getCoderRegistry()),
+                        doFnInvoker.invokeGetWatermarkEstimatorStateCoder(
```

Comment:
Just curious what if someone has SDF without provide WatermarkEstimator?

Reply:
The default watermark estimator state is `void` and always returns the min timestamp.

See:
https://github.com/apache/beam/blob/7310ec22078d50ddf9ccec054bf0e7aad33a7e7f/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java#L665

Reply:
Thanks for the explanation!